crawl-ref/source/dungeon.cc

```diff
@@ -5875,14 +5875,19 @@ static void _place_specific_trap(const coord_def& where, trap_spec* spec,
                 env.placing_vault.c_str());
     }
 
+    // find an appropriate trap for TRAP_RANDOM
     while (spec_type >= NUM_TRAPS
 #if TAG_MAJOR_VERSION == 34
            || spec_type == TRAP_DART || spec_type == TRAP_GAS
            || spec_type == TRAP_SHADOW || spec_type == TRAP_SHADOW_DORMANT
 #endif
            || !is_valid_shaft_level() && spec_type == TRAP_SHAFT)
     {
-        spec_type = static_cast<trap_type>(random2(TRAP_MAX_REGULAR + 1));
+        do
+        {
+            spec_type = static_cast<trap_type>(random2(NUM_TRAPS));
+        }
+        while (!is_regular_trap(spec_type))
     }
```

Comment:
Having a `do...while` loop inside a `while` loop like this seems really ugly and inefficient. Is there a better way to do this so that only the `is_regular_trap` condition is checked at the end of every loop whereas all of the other conditions are checked at the start?

Reply:
This loop only runs for TRAP_RANDOM. I'd replace the `spec_type >= NUM_TRAPS` clause by wrapping the whole loop in an explicit `if (spec_type == TRAP_RANDOM)` block and moving the remainder of the conditions to the single `do...while`

Reply:
Ah yea that's neater. I missed the fact that shafts are handled in an earlier `if` and so this loop doesn't run for shafts on the final floors of branches.